### ISSUE TITLE
Generative Adversarial Imitation Learning

### ISSUE THREAD

**Issue by magnusja** [tensorforce/tensorforce#302]
Hey guys,

recently the paper Generative Adversarial Imitation Learning [1] and some follow up papers [2,...] have gained attraction. I am thinking about implementing GAIL using TensorForce.
The main difference between GAIL and usual RL is that one does not have a simulator to get reward feedback from. Instead, you have expert demonstrations (trajectories of state action pairs) and you are using a Discriminator (similar to GANs) to distinguish between real demonstrated trajectories by the expert and fake demonstrations generated by the policy network. You can then use the output of the discriminator as feedback and optimize the policy network using TRPO or some other PG method.

Algorithm sketch from the paper[1]
![image](https://user-images.githubusercontent.com/1659457/34960702-cfbcc6ec-f9f0-11e7-968a-2cd2f4f20d82.png)

Do you guys think it makes sense to directly include that in TensorForce?
How would you approach the implementation? I think the easiest approach would be to create a new Discriminator environment. The only thing is that I need to somehow get the training of the discriminator into the TensorFlow Graph? How would you approach this considering the current TensorForce architecture?

[1] https://arxiv.org/pdf/1606.03476.pdf
[2] https://arxiv.org/abs/1703.08840
[3] https://arxiv.org/abs/1701.06699
[4] https://arxiv.org/abs/1707.02747
[5] https://arxiv.org/pdf/1611.03852.pdf

**Comment by michaelschaarschmidt** [comment]
Hey,

1. This is definitely cool and of interest, as learning from demonstration seems to be the way forward to manage some of the instabilities of current algorithms.

2. A discriminator environment sounds like a good idea to start off with but some things to consider.
Moving this into the TensorFlow graph might be tricky yes, and we will have to have a bit of a discussion on this:

If it's part of the main TensorFlow graph, it's hard to see as an environment, because we try to separate environment and model. Can we put the discriminator as an extra model that extends another model? In general, this seems like a generic type of a new architecture we would want to support, and we would want to be able to plug any two combinations into discriminator and whatever optimizes the policy.

Having it as an environment seems easier but then there would maybe two separate graphs?

**Comment by AlexKuhnle** [comment]
Hi @magnusja,

I agree with Michael that this sounds like an interesting feature, and I think we can work towards an implementation in TensorForce.

First a clarification: Why do you see the discriminator as an 'environment'? Do you mean an environment in the TensorForce sense (i.e. a subclass of the `Environment` class)? If you consider the discriminator a separate agent with its own 'expert environment' or something, I think this is probably not the best way to integrate this.

Here my version: The (not yet existing, but relatively obviously at some point required) component where this fits best is a `RewardEstimator` (or whatever it would be called) that, given state/action/etc gives the model the reward signal. In this case the reward estimator has its own discriminator network and memory with expert trajectories, which are trained separately from the model, but at the same time (or not). Does this make sense, or do you see any problems? Happy to give input (in fact I would prefer giving some input :-) if you want to have a go at implementing it (e.g. via Gitter private chat or mail).  :+1:

**Comment by magnusja** [comment]
Hey guys,

@AlexKuhnle 
Yes, I mean the discriminator to be a subclass of the `Environment` class. The main reason why I would see as an environment is because it would be very easy to implement with the current architecture. I would assume that I have some kind of dataset with states and demonstrated actions. The discriminator environment loads these, trains itself and then acts as an environment for the agent. It takes in actions from the agent and emits reward and also next states. Next states would then come from another simulator (which can only simulate state transitions but no rewards), for instance, depending on the problem.

Regarding the `RewardEstimator`: So you would like to split emitting rewards and emitting next states into two separate parts, which are one at the moment (like OpenAI gym). Or how would the `RewardEestimator` look like in the normal RL case?

@michaelschaarschmidt 
So I think it is not a problem having two separate graphs because in the normal GAIL case they do not need to be connected. There is another paper called Model-based Adversarial Imitation Learning[1], where they backpropagate from the discriminator into the generator/policy. In that case, they would need to be connected. But I think we can move forward with the simpler case for now?

[1] https://arxiv.org/pdf/1612.02179.pdf

**Comment by AlexKuhnle** [comment]
Ah, I see what you mean. Then my comment is suggesting a different design: The environment stays the same, so still outputs some reward, but the `reward_estimator` can be configured to GAIL (instead of, say, default none for using environment reward), which returns its own independent reward signal. So really just a shift of the functionality from within the environment to within the model as a sub-component. Some thoughts on why I suggest this:

(a) the `Environment` class is optional anyway and definitely not supposed to contain model logic,
(b) some kind of `RewardEstimator` object seems useful/necessary for different models and hence might be a good idea,
(c) the discriminator and its optimization should probably be coupled with processes within the model (e.g. optimization),
(d) the latter case of backpropagating through both parts seems more doable that way.

What do you think?

**Comment by magnusja** [comment]
So you suggest that the reward estimator is part of the model? Or where would you hook that into the pipeline?

**Comment by AlexKuhnle** [comment]
Yes, I imagine it something like a 'reward preprocessor', see [here](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/models/model.py#L829). GAIL would just discard the external reward signal and instead return its own. Of course this part needs to be extended quite a bit, it can't just literally be implemented as a subclass to `Preprocessor`, but just to give an idea of how I think it would integrate well. I could probably give you a abstract class interface and you could see whether this makes sense. Definitely also worth building this on the soon-to-be-pre-released (weekend probably) branch for the memory update, since some bits in the model get moved around.

Regarding your second point, you're indeed right that the agent assumes unbatched inputs, even though the model can handle batched ones as well. This can probably best be addressed in the aforementioned 'memory branch', since changes touch on these parts, too. Thanks for spotting this inconsistency.

**Comment by AlexKuhnle** [comment]
I can try to come up with a sketch later, busy this weekend getting the memory branch in shape and pushed. You can check it out now. However, it's not possible to pass tensors, but have a look at DQFD and how a separate demonstration memory is handled there. I don't expect we will extend the interface to allow for TensorFlow input anytime soon, if at all. We think of the TensorForce API as hiding all the necessary TensorFlow behind an agnostic and easy-to-use interface, so a somewhat different focus. But I'm confident we can integrate GAIL without this feature.

**Comment by AlexKuhnle** [comment]
Added to roadmap.
